core/src/apps/cardano/address.py

```diff
@@ -1,145 +1,344 @@
+        return bech32.encode(hrp, address)
+    else:
+        raise ValueError("Invalid address")
+
+
+def derive_human_readable_address(
+    keychain: seed.Keychain,
+    parameters: CardanoAddressParametersType,
+    protocol_magic: int,
+    network_id: int,
+) -> str:
+    address = derive_address_bytes(keychain, parameters, protocol_magic, network_id)
+    return get_human_readable_address(address, network_id)
+
+
+def derive_address_bytes(
+    keychain: seed.Keychain,
+    parameters: CardanoAddressParametersType,
+    protocol_magic: int,
+    network_id: int,
+) -> bytes:
+    is_byron_address = parameters.address_type == CardanoAddressType.BYRON
+
+    if is_byron_address:
+        address = _derive_byron_address(
+            keychain, parameters.spending_key_path, protocol_magic
+        )
+    else:
+        address = _derive_shelley_address(keychain, parameters, network_id)
+
+    return address
 
-        if not isinstance(address_protocol_magic, int):
-            raise INVALID_ADDRESS
 
-        if address_protocol_magic != protocol_magic:
-            raise NETWORK_MISMATCH
+def _derive_byron_address(
+    keychain: seed.Keychain, path: List[int], protocol_magic: int
+) -> bytes:
+    if not is_byron_path(path):
+        raise wire.DataError("Invalid path for byron address!")
 
+    address = derive_byron_address(keychain, path, protocol_magic)
+    return address
 
-def validate_full_path(path: list) -> bool:
+
+def _derive_shelley_address(
+    keychain: seed.Keychain, parameters: CardanoAddressParametersType, network_id: int,
+) -> bytes:
+    if not is_shelley_path(parameters.spending_key_path):
+        raise wire.DataError("Invalid path for shelley address!")
+
+    address_header = _create_address_header(parameters.address_type, network_id)
+
+    if parameters.address_type == CardanoAddressType.BASE:
+        address = _derive_base_address(
+            keychain,
+            address_header,
+            parameters.spending_key_path,
+            parameters.staking_key_path,
+            parameters.staking_key_hash,
+        )
+    elif parameters.address_type == CardanoAddressType.ENTERPRISE:
+        address = _derive_enterprise_address(
+            keychain, address_header, parameters.spending_key_path
+        )
+    elif parameters.address_type == CardanoAddressType.POINTER:
+        address = _derive_pointer_address(
+            keychain,
+            address_header,
+            parameters.spending_key_path,
+            parameters.certificate_pointer,
+        )
+    elif parameters.address_type == CardanoAddressType.REWARD:
+        address = _derive_reward_address(
+            keychain, address_header, parameters.spending_key_path
+        )
+    else:
+        raise ValueError("Invalid address type '%s'!" % parameters.address_type)
+
+    return address
+
+
+def _create_address_header(address_type: int, network_id: int) -> bytes:
```

Comment:
Do you mean `CardanoAddressType`? I've changed it to that [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/76bc4dcc3a298bc190d0cd4263fae7ed1d998d73), since I don't know if `EnumTypeCardanoAddressType` is a thing.

Reply:
no, it's actually `EnumTypeCardanoAddressType` and it's importable from `CardanoAddressParametersType`

Reply:
that applies to most places where you use `CardanoAddressType` as a type. Although it sounds like it should be, it's not a valid mypy type :)

Reply:
Yes, sorry, I noticed there's some more usages I should replace. Done [here](https://github.com/trezor/trezor-firmware/pull/1112/commits/43c332ca1c320bf078669f847154ecc84c68122f).